### ISSUE TITLE
No One Lives Forver 2 (NOLF2)

### ISSUE THREAD

**Issue by UserGit-2021** [kcat/dsoal#61]
Experimented
Versions:
DSOAL - Don't uninitialize COM while an IMMDevice is open
OpenAL Soft - Don't pass an EaxCall to initialize_fx_slots 
At this level
Dialog work 3d eax

![изображение](https://user-images.githubusercontent.com/91757058/179464754-bd5d7358-8b86-4549-b7a6-5ae200fedad6.png)

At the beginning of the level
Dialog doesn't work eax

![изображение](https://user-images.githubusercontent.com/91757058/179464979-0d047aa7-209a-4621-8fdc-6fd8f92e5ddf.png)

This is where EAX conversations work.
Probably the conversations of the enemies are working.
NPC story dialogues do not work.

![изображение](https://user-images.githubusercontent.com/91757058/179465189-b2c2cc86-28d2-4e26-b921-6cc3a729c0f3.png)

Interesting observation! If the bird is hurt, the dialogue sound will work EAX.🤔 🤔 🤔 

![изображение](https://user-images.githubusercontent.com/91757058/179465483-cf50b237-4878-4267-9731-119ecaab6a82.png)

With DSOAL music doesn't work.
ALchemy works echo dialog
DSOAL doesn't works echo dialog

P.S. dsoal_log_error does not create.

**Comment by UserGit-2021** [comment]
I'm sorry that I'm bad at describing problems.

**Comment by UserGit-2021** [comment]
Video test NPC dialog. Music not working with DSOAL.
DSOAL
https://drive.google.com/file/d/1Q-D-JHchvU4Sb0gs4mP2riRTg_962TE7/view?usp=sharing

ALchemy
https://drive.google.com/file/d/14gTBlLC-SlXYa05jplU5R72jr3MM2uxi/view?usp=sharing

**Comment by ThreeDeeJay** [comment]
I can confirm shooting that robot bird near the first level makes its dialog get spatialized. Kinda bizarre.
Also, it seems the listener orientation doesn't rotate vertically with the camera, tho perhaps that might just be an oversight in the game's DirectSound3D implementation. Here's some footage: https://youtu.be/7sYlcbH_91g?t=367

**Comment by UserGit-2021** [comment]
Contract J.A.C.K.
[dsoal_error.txt](https://github.com/kcat/dsoal/files/9182567/dsoal_error.txt)

**Comment by ThreeDeeJay** [comment]
> I can confirm shooting that robot bird near the first level makes its dialog get spatialized. Kinda bizarre.
By the way, this also happens in XP with real hardware and just like ALchemy, NPC dialog is non-spatial stereo (https://youtu.be/tMG-aa55Ac8), while in DSOAL it sounds like virtual front center speaker, not sure if that's even relevant but the behavior could provide a clue in case all dialog can be forced to be spatialized at all times, tho I wouldn't be surprised if the game was just hardcoded to behave like this.

**Comment by UserGit-2021** [comment]
![изображение](https://user-images.githubusercontent.com/91757058/230740740-1e8d3389-b2cf-447c-9d12-1fcd80f1c5a4.png)

**Comment by kcat** [comment]
That seems to be a result of the same problem in issue #71. DirectMusic uses an undocumented internal interface by querying GUID `2a8af120-e9de-4132-aaa5-4bdda5f325b8` from the device. DSOAL errors on that since it doesn't know what it is, which causes DirectMusic to fail. That interface will need to be implemented for it to work.

**Comment by jeremywh7** [comment]
Searching for "2a8af120-e9de-4132-aaa5-4bdda5f325b8" brings up some interesting results, e.g. from [IndirectSound v0.14](https://www.indirectsound.com/downloads.html):
---------------------
Use Microsoft's COM object when IID {2A8AF120-E9DE-4132-AAA5-4BDDA5F325B8} is queried
 -This can at least partially fix DirectMusic in some games, but it may also cause problems of its own (in No One Lives Forever 2, for example, this changes restores missing music but there can still be crashes inside DirectMusic which may or may not be related)
 -DirectMusic's IDirectMusicPerformance8::CreateAudioPath() can call DirectSound and query it for an undocumented IID. IndirectSound will now recognize this IID, pass it on directly to Microsoft's official DirectSound, and return the resulting COM object to the application. This means, however, that DirectMusic is using a pure DirectSound COM object that is not controlled by IndirectSound in any way (specifically, the COM object that DirectMusic has does not support hardware acceleration).

**Comment by kcat** [comment]
Yes, that was mentioned in that other issue. What's more interesting is this from Hourglass

https://github.com/google-code-export/hourglass-win32/blob/master/src/wintasee/hooks/soundhooks.cpp#L2032

I have no idea what this is or where it came from, but it at least gives a start for the kind of interface it's expecting. I have no idea if I'll actually be able to implement it satisfactorily since there's even more layers to it, but it's at least a starting point.

**Comment by mirh** [comment]
I really don't think that's the most interesting thing from what I figured out...
(in fact hourglass may be as well completely off the mark given `IID_IDirectSoundSinkFactory` doesn't seem to exist anywhere else on the net, including the [symbols](http://lise.pnfsoftware.com/winpdb/A98690D18A3AF9C6604DD8FEA198030F719612134F4D78D77FF08C8010671969-dmusic.html))

**Comment by kcat** [comment]
> I really don't think that's the most interesting thing from what I figured out...
(in fact hourglass may be as well completely off the mark given `IID_IDirectSoundSinkFactory` doesn't seem to exist anywhere else on the net, including the [symbols](http://lise.pnfsoftware.com/winpdb/A98690D18A3AF9C6604DD8FEA198030F719612134F4D78D77FF08C8010671969-dmusic.html))

`IID_IDirectSoundSinkFactory` needn't exist as its own symbol in dmusic, since it's an internal DSound interface the GUID can be stored as an unmarked static variable, while `IDirectSoundSinkFactory` is a basic vtable struct without type information to store.

But if nothing else, it gives me something to try. If it turns out to be off the mark, then we're no better or worse off. If it seems right, then we may have something to build on. Given [the comments made](https://github.com/google-code-export/hourglass-win32/blob/master/src/wintasee/hooks/soundhooks.cpp#L2118), it seems likely they did get somewhere with it.

**Comment by mirh** [comment]
GUIDs can be stored and called with whatever unknown unnamed variable you want... 
But since we are actually giving them names here, absence of evidence is kinda evidence of absence. 

And while I have the confidence in my RE skills below the knee, we also seem to know that nitsuja guy was really "???" in many places of the code, and that they did screw up already by mixing the `IID_IDirectSoundSynthSink` value with `IID_IDirectSoundSinkSync`.

**Comment by UserGit-2021** [comment]
UPD: The music still doesn't work, but the conversations already work in 3D, without injuring the bird.

![Image](https://github.com/user-attachments/assets/d6f73fb2-2878-4742-8bb7-6bcc8a0c3dc2)

![Image](https://github.com/user-attachments/assets/5f75a8d3-351e-4060-a942-f95aacf99c9c)

**Comment by UserGit-2021** [comment]
DSOAL r626@b27f1025 + OpenAL Soft r9603@30eb6c3c

Only music doesn't work

[alsoft_error.txt](https://github.com/user-attachments/files/18585604/alsoft_error.txt)
[dsoal_error.txt](https://github.com/user-attachments/files/18585603/dsoal_error.txt)

![Image](https://github.com/user-attachments/assets/24d4aa4e-61ef-4336-931a-9ef41fd4d48e)
